In [1]:
import unittest

class TestStringMethods(unittest.TestCase):

    def test_upper(self):
        self.assertEqual('foo'.upper(), 'FOO')

    def test_isupper(self):
        self.assertTrue('FOO'.isupper())
        self.assertFalse('Foo'.isupper())

    def test_split(self):
        s = 'hello world'
        self.assertEqual(s.split(), ['hello', 'world'])
        # check that s.split fails when the separator is not a string
        with self.assertRaises(TypeError):
            s.split(2)

if __name__ == '__main__':
    unittest.main()

E
ERROR: /root/ (unittest.loader._FailedTest./root/)
----------------------------------------------------------------------
AttributeError: module '__main__' has no attribute '/root/'

----------------------------------------------------------------------
Ran 1 test in 0.003s

FAILED (errors=1)


SystemExit: True

/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py:3561: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [3]:
!python -m unittest test_module1 test_module2
!python -m unittest test_module.TestClass
!python -m unittest test_module.TestClass.test_method

EE
ERROR: test_module1 (unittest.loader._FailedTest.test_module1)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_module1
Traceback (most recent call last):
  File "/usr/lib/python3.11/unittest/loader.py", line 162, in loadTestsFromName
    module = __import__(module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'test_module1'


ERROR: test_module2 (unittest.loader._FailedTest.test_module2)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_module2
Traceback (most recent call last):
  File "/usr/lib/python3.11/unittest/loader.py", line 162, in loadTestsFromName
    module = __import__(module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'test_module2'


----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (errors=2)
E
ERROR: test_modul

In [4]:
def greet(name, message="Hello", *args, **kwargs):
    """Prints a greetings message"""
    print(message, name)
    for arg in args:
        print(arg)
    for key, value in kwargs.items():
        print(f"{key}: {value}")

# Example usage:
greet("John", "Hi", "Nice to meet you", age=30, country="USA")


Hi John
Nice to meet you
age: 30
country: USA


In [5]:
def function_name(parameters):
    # function body
    pass


In [6]:
def add_numbers(num1, num2):
    """Returns the sum of two numbers"""
    total = num1 + num2
    return total

# Example usage:
result = add_numbers(5, 7)
print(result)  # Output: 12


12


In [9]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = gesamt_groesse // teile_groesse_mb

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = (i + 1) * teile_groesse_mb
        teile.append(daten.iloc[start:end])

    # Restliche Daten
    rest = daten.iloc[anzahl_teile * teile_groesse_mb:]
    if not rest.empty:
        teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["datei1.parquet", "datei2.parquet", "datei3.parquet"]
suche_text = "Beispieltext"
teile_groesse_mb = 50  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

Fehler beim Lesen der Datei datei1.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei1.parquet
Fehler beim Lesen der Datei datei2.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei2.parquet
Fehler beim Lesen der Datei datei3.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei3.parquet


ValueError: No objects to concatenate

In [8]:
!pip install dask[dataframe]


INFO: pip is looking at multiple versions of dask-expr to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.2/243.2 kB 3.9 MB/s eta 0:00:00


In [10]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = gesamt_groesse // teile_groesse_mb

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = (i + 1) * teile_groesse_mb
        teile.append(daten.iloc[start:end])

    # Restliche Daten
    rest = daten.iloc[anzahl_teile * teile_groesse_mb:]
    if not rest.empty:
        teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["datei1.parquet", "datei2.parquet", "datei3.parquet"]
suche_text = "Beispieltext"
teile_groesse_mb = 50  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

Fehler beim Lesen der Datei datei1.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei1.parquet
Fehler beim Lesen der Datei datei2.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei2.parquet
Fehler beim Lesen der Datei datei3.parquet: An error occurred while calling the read_parquet method registered to the pandas backend.
Original Message: /content/datei3.parquet
Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.
Ähnliche Daten: []


In [11]:
!wget https://huggingface.co/datasets/Salesforce/wikitext/resolve/main/wikitext-2-v1/train-00000-of-00001.parquet

--2025-01-24 18:51:39--  https://huggingface.co/datasets/Salesforce/wikitext/resolve/main/wikitext-2-v1/train-00000-of-00001.parquet
Resolving huggingface.co (huggingface.co)... 18.244.202.118, 18.244.202.73, 18.244.202.68, ...
Connecting to huggingface.co (huggingface.co)|18.244.202.118|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cdn-lfs.hf.co/datasets/wikitext/dfc27e4360c639dc1fba1e403bfffd53af4a5c75d5363b5724d49bf12d07cce6?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27train-00000-of-00001.parquet%3B+filename%3D%22train-00000-of-00001.parquet%22%3B&Expires=1737748299&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTczNzc0ODI5OX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5oZi5jby9kYXRhc2V0cy93aWtpdGV4dC9kZmMyN2U0MzYwYzYzOWRjMWZiYTFlNDAzYmZmZmQ1M2FmNGE1Yzc1ZDUzNjNiNTcyNGQ0OWJmMTJkMDdjY2U2P3Jlc3BvbnNlLWNvbnRlbnQtZGlzcG9zaXRpb249KiJ9XX0_&Signature=PI98vH4meV4GfXYDpDaUmf8AYvpeojXv%7EZTLmK6nJgdRx

In [13]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = gesamt_groesse // teile_groesse_mb

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = (i + 1) * teile_groesse_mb
        teile.append(daten.iloc[start:end])

    # Restliche Daten
    rest = daten.iloc[anzahl_teile * teile_groesse_mb:]
    if not rest.empty:
        teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["train-00000-of-00001.parquet"]
suche_text = "Beispieltext"
teile_groesse_mb = 2  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

ValueError: Found array with 0 sample(s) (shape=(0, 1)) while a minimum of 1 is required by TfidfTransformer.

شغال

In [14]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = int(gesamt_groesse // teile_groesse_mb)  # Ensure integer division

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = min((i + 1) * teile_groesse_mb, len(daten))  # Limit end index
        # Use loc instead of iloc to slice by row label rather than row number
        teil = daten.loc[start:end]  # Use loc to slice by row label
        if not teil.empty:  # Only add non-empty chunks
            teile.append(teil)

    # Restliche Daten (handle if the last chunk is smaller than teile_groesse_mb)
    if len(daten) > anzahl_teile * teile_groesse_mb:
        rest = daten.loc[anzahl_teile * teile_groesse_mb:]
        if not rest.empty:
            teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["/content/train-00000-of-00001.parquet"]
suche_text = "Beispieltext"
teile_groesse_mb = 2  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

Ähnliche Daten: [12240, 12241, 12242, 12236, 36717]



شرح الكود باللغة العربية:
يقوم الكود بتحليل ملفات البيانات بتنسيق "parquet" للبحث عن نصوص مشابهة لنص معين يُحدده المستخدم. إليك شرح مبسط لكل جزء من الكود:

1. استيراد المكتبات:

pandas: لمتلفة البيانات وتحليلها باستخدام إطار بيانات يُسمى "DataFrame".
dask.dataframe: للتعامل مع مجموعات البيانات الكبيرة التي قد لا تتسع في الذاكرة.
TfidfVectorizer: لتحويل النصوص إلى متجهات رقمية باستخدام خوارزمية TF-IDF، والتي تساعد في تحديد أهمية الكلمات في النص.
cosine_similarity: لحساب تشابه النصوص بناءً على متجهاتها الرقمية.
os: للتفاعل مع نظام التشغيل، مثل التعامل مع مسارات الملفات.
2. دالة lesen_parquet_datei:

تقرأ ملف "parquet" وتحوله إلى إطار بيانات "DataFrame".
تستخدم مكتبة "Dask" للقراءة بكفاءة، ثم تُحوّل النتيجة إلى "DataFrame" باستخدام .compute().
في حال حدوث خطأ، تطبع رسالة خطأ وتُعيد القيمة "None".
3. دالة suche_aehnliche_daten:

جوهر عملية البحث عن النصوص المشابهة.
تستقبل إطار بيانات "DataFrame" ونص البحث كمدخلات.
تُهيئ TfidfVectorizer لتحويل النصوص إلى متجهات رقمية، مع إزالة الكلمات الشائعة باللغة الإنجليزية مثل "the" و "a" و "is" باستخدام stop_words="english".
تُحوّل نص البحث إلى متجه رقمي باستخدام fit_transform.
تُحوّل عمود "text" في إطار البيانات إلى متجهات رقمية باستخدام transform.
تحسب تشابه النصوص باستخدام cosine_similarity.
تُعيد مؤشرات النصوص الخمسة الأكثر تشابهًا مع نص البحث.
4. دالة teilen_daten:

تُقسّم إطار البيانات إلى أجزاء أصغر لتسهيل التعامل مع مجموعات البيانات الكبيرة.
تحسب حجم كل جزء بناءً على teile_groesse_mb وحجم الذاكرة المُستخدم من إطار البيانات.
تستخدم loc للتقسيم بناءً على مُسميات الصفوف.
تُعالج حالات الجزء الأخير الذي قد يكون أصغر من الحجم المُحدد.
5. دالة haupt_funktion:

الدالة الرئيسية التي تُدير عملية قراءة البيانات، وتقسيمها، والبحث عن نصوص مشابهة.
تستقبل أسماء الملفات، نص البحث، وحجم الجزء كمدخلات.
تقرأ ملفات "parquet" ، وتدمجها في إطار بيانات واحد.
تُقسّم إطار البيانات إلى أجزاء.
تستدعي دالة suche_aehnliche_daten على كل جزء للبحث عن النصوص المشابهة.
تجمع النتائج من جميع الأجزاء وتُعيدها.
6. استخدام الكود:

يُحدد أسماء الملفات، نص البحث، وحجم الجزء.
يستدعي دالة haupt_funktion للحصول على النتائج.
يطبع النتائج.
معنى النتيجة:
عندما تحصل على ناتج مثل "Ähnliche Daten: [12240, 12241, 12242, 12236, 36717]"، فهذا يعني أن الكود وجد خمسة نصوص في ملفات "parquet" تشبه نص البحث "Beispieltext".

الأرقام في القائمة هي مُؤشرات (أرقام الصفوف) لهذه النصوص المشابهة داخل إطار البيانات.
بمعنى آخر، الصفوف ذات الأرقام 12240، 12241، 12242، 12236، و 36717 تحتوي على نصوص تشبه "Beispieltext" بناءً على حساب تشابه النصوص باستخدام خوارزمية TF-IDF.
ببساطة: لقد بحثت عن نص مشابه لـ "Beispieltext" في بياناتك، ووجد الكود خمسة صفوف في بياناتك تُطابق بحثك بشكل كبير. الأرقام في الناتج تُشير إلى مواقع هذه الصفوف في بياناتك.



شغال

In [15]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = int(gesamt_groesse // teile_groesse_mb)  # Ensure integer division

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = min((i + 1) * teile_groesse_mb, len(daten))  # Limit end index
        # Use loc instead of iloc to slice by row label rather than row number
        teil = daten.loc[start:end]  # Use loc to slice by row label
        if not teil.empty:  # Only add non-empty chunks
            teile.append(teil)

    # Restliche Daten (handle if the last chunk is smaller than teile_groesse_mb)
    if len(daten) > anzahl_teile * teile_groesse_mb:
        rest = daten.loc[anzahl_teile * teile_groesse_mb:]
        if not rest.empty:
            teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["/content/train-00000-of-00001.parquet"]
suche_text = " Valkyira Chronicles"
teile_groesse_mb = 2  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

Ähnliche Daten: [26, 10068, 13322, 21, 46]


In [ ]:
لتغيير النص المراد البحث عنه، ما عليك سوى تعديل قيمة هذا المتغير.

ستجد السطر التالي في الكود:


suche_text = "Beispieltext"

شغال

In [16]:
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    idx = aehnlichkeit.argsort()[0][-5:]
    return daten.iloc[idx]

def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        gefunden = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.append(gefunden)

    return aehnliche_daten

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
for i, gefunden in enumerate(aehnliche_daten):
    print(f"Ähnliche Daten {i+1}:")
    print(gefunden)
    print("\n")

Ähnliche Daten 1:
                                                    text
26      The music was composed by Hitoshi Sakimoto , ...
10068   Harold Innis 's interest in the relationship ...
13322   Some of the original stories were well @-@ re...
21      Concept work for Valkyria Chronicles III bega...
46      Valkyria Chronicles 3 was adapted into a two ...




شغال1

In [1]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = int(gesamt_groesse // teile_groesse_mb)  # Ensure integer division

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = min((i + 1) * teile_groesse_mb, len(daten))  # Limit end index
        # Use loc instead of iloc to slice by row label rather than row number
        teil = daten.loc[start:end]  # Use loc to slice by row label
        if not teil.empty:  # Only add non-empty chunks
            teile.append(teil)

    # Restliche Daten (handle if the last chunk is smaller than teile_groesse_mb)
    if len(daten) > anzahl_teile * teile_groesse_mb:
        rest = daten.loc[anzahl_teile * teile_groesse_mb:]
        if not rest.empty:
            teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["/content/train-00000-of-00001.parquet"]
suche_text = " Valkyira Chronicles"
teile_groesse_mb = 2  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)

Ähnliche Daten: [26, 10068, 13322, 21, 46]


شغال2

In [2]:
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    idx = aehnlichkeit.argsort()[0][-5:]
    return daten.iloc[idx]

def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        gefunden = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.append(gefunden)

    return aehnliche_daten

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
for i, gefunden in enumerate(aehnliche_daten):
    print(f"Ähnliche Daten {i+1}:")
    print(gefunden)
    print("\n")

Ähnliche Daten 1:
                                                    text
26      The music was composed by Hitoshi Sakimoto , ...
10068   Harold Innis 's interest in the relationship ...
13322   Some of the original stories were well @-@ re...
21      Concept work for Valkyria Chronicles III bega...
46      Valkyria Chronicles 3 was adapted into a two ...




شغال بحث عن المتشابهات فى ملف داتاسيت
https://huggingface.co/datasets/Salesforce/wikitext/viewer

In [3]:
import pandas as pd
import dask.dataframe as dd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import os

# Funktion zum Lesen von parquet-Dateien
def lesen_parquet_datei(dateiname):
    try:
        # Mit Dask lesen wir die Datei in einem Datenrahmen
        df = dd.read_parquet(dateiname)
        return df.compute()
    except Exception as e:
        print(f"Fehler beim Lesen der Datei {dateiname}: {e}")
        return None

# Funktion zur Suche nach ähnlichen Daten
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    return aehnlichkeit.argsort()[0][-5:]

# Funktion zum Teilen der Daten in Teile
def teilen_daten(daten, teile_groesse):
    # Die Größe jeder Teilung
    teile_groesse_mb = teile_groesse * 1024 * 1024
    # Die Gesamtgröße der Daten
    gesamt_groesse = daten.memory_usage().sum()
    # Die Anzahl der Teile
    anzahl_teile = int(gesamt_groesse // teile_groesse_mb)  # Ensure integer division

    # Teile der Daten erstellen
    teile = []
    for i in range(anzahl_teile):
        start = i * teile_groesse_mb
        end = min((i + 1) * teile_groesse_mb, len(daten))  # Limit end index
        # Use loc instead of iloc to slice by row label rather than row number
        teil = daten.loc[start:end]  # Use loc to slice by row label
        if not teil.empty:  # Only add non-empty chunks
            teile.append(teil)

    # Restliche Daten (handle if the last chunk is smaller than teile_groesse_mb)
    if len(daten) > anzahl_teile * teile_groesse_mb:
        rest = daten.loc[anzahl_teile * teile_groesse_mb:]
        if not rest.empty:
            teile.append(rest)

    return teile

# Haupt-Funktion
def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        aehnlichkeit = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.extend(aehnlichkeit)

    return aehnliche_daten

# Verwendung
dateinamen = ["/content/train-00000-of-00001.parquet"]
suche_text = " Valkyira Chronicles"
teile_groesse_mb = 2  # Größe jeder Teilung in MB

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
print("Ähnliche Daten:", aehnliche_daten)
def suche_aehnliche_daten(daten, suche_text):
    # TF-IDF-Vektorisierung
    vektorisierung = TfidfVectorizer(stop_words="english")
    # Vektorisierung der Suchtextes
    suche_vektor = vektorisierung.fit_transform([suche_text])
    # Vektorisierung der Daten
    daten_vektor = vektorisierung.transform(daten["text"])

    # Berechnung der Ähnlichkeit mithilfe der Kosinus-Ähnlichkeit
    aehnlichkeit = cosine_similarity(suche_vektor, daten_vektor)

    # Rückgabe der Top-5 ähnlichsten Ergebnisse
    idx = aehnlichkeit.argsort()[0][-5:]
    return daten.iloc[idx]

def haupt_funktion(dateinamen, suche_text, teile_groesse_mb=100):
    # Lesen der Dateien
    datenframes = []
    for dateiname in dateinamen:
        daten = lesen_parquet_datei(dateiname)
        if daten is not None:
            datenframes.append(daten)

    # Check if any dataframes were read successfully
    if not datenframes:  # If datenframes is empty
        print("Keine Daten konnten gelesen werden. Überprüfen Sie die Dateinamen und Pfade.")
        return []  # Return an empty list to avoid the ValueError

    # Verkettung der Daten
    daten = pd.concat(datenframes)

    # Teilen der Daten in Teile
    teile = teilen_daten(daten, teile_groesse_mb)

    # Suche nach ähnlichen Daten
    aehnliche_daten = []
    for teil in teile:
        gefunden = suche_aehnliche_daten(teil, suche_text)
        aehnliche_daten.append(gefunden)

    return aehnliche_daten

aehnliche_daten = haupt_funktion(dateinamen, suche_text, teile_groesse_mb)
for i, gefunden in enumerate(aehnliche_daten):
    print(f"Ähnliche Daten {i+1}:")
    print(gefunden)
    print("\n")

Ähnliche Daten: [26, 10068, 13322, 21, 46]
Ähnliche Daten 1:
                                                    text
26      The music was composed by Hitoshi Sakimoto , ...
10068   Harold Innis 's interest in the relationship ...
13322   Some of the original stories were well @-@ re...
21      Concept work for Valkyria Chronicles III bega...
46      Valkyria Chronicles 3 was adapted into a two ...




https://huggingface.co/datasets/Salesforce/wikitext/viewer?row=32

The game takes place during the Second Europan War . Gallian Army Squad 422 , also known as " The Nameless " , are a penal military unit composed of criminals , foreign deserters , and military offenders whose real names are erased from the records and thereon officially referred to by numbers . Ordered by the Gallian military to perform the most dangerous missions that the Regular Army and Militia will not do , they are nevertheless up to the task , exemplified by their motto , Altaha Abilia , meaning " Always Ready . " The three main characters are No.7 Kurt Irving , an army officer falsely accused of treason who wishes to redeem himself ; Ace No.1 Imca , a female Darcsen heavy weapons specialist who seeks revenge against the Valkyria who destroyed her home ; and No.13 Riela Marcellis , a seemingly jinxed young woman who is unknowingly a descendant of the Valkyria . Together with their fellow squad members , these three are tasked to fight against a mysterious Imperial unit known as Calamity Raven , consisting of mostly Darcsen soldiers .
As the Nameless officially do not exist , the upper echelons of the Gallian Army exploit the concept of plausible deniability in order to send them on missions that would otherwise make Gallia lose face in the war . While at times this works to their advantage , such as a successful incursion into Imperial territory , other orders cause certain members of the 422nd great distress . One such member , Gusurg , becomes so enraged that he abandons his post and defects into the ranks of Calamity Raven , attached to the ideal of Darcsen independence proposed by their leader , Dahau . At the same time , elements within Gallian Army Command move to erase the Nameless in order to protect their own interests . Hounded by both allies and enemies , and combined with the presence of a traitor within their ranks , the 422nd desperately move to keep themselves alive while at the same time fight to help the Gallian war effort . This continues until the Nameless 's commanding officer , Ramsey Crowe , who had been kept under house arrest , is escorted to the capital city of Randgriz in order to present evidence exonerating the weary soldiers and expose the real traitor , the Gallian General that had accused Kurt of Treason .
Partly due to these events , and partly due to the major losses in manpower Gallia suffers towards the end of the war with the Empire , the Nameless are offered a formal position as a squad in the Gallian Army rather than serve as an anonymous shadow force . This is short @-@ lived , however , as following Maximilian 's defeat , Dahau and Calamity Raven move to activate an ancient Valkyrian super weapon within the Empire , kept secret by their benefactor . Without the support of Maximilian or the chance to prove themselves in the war with Gallia , it is Dahau 's last trump card in creating a new Darcsen nation . As an armed Gallian force invading the Empire just following the two nations ' cease @-@ fire would certainly wreck their newfound peace , Kurt decides to once again make his squad the Nameless , asking Crowe to list himself and all under his command as killed @-@ in @-@ action . Now owing allegiance to none other than themselves , the 422nd confronts Dahau and destroys the Valkyrian weapon . Each member then goes their separate ways in order to begin their lives anew .
= = Development = =
Concept work for Valkyria Chronicles III began after development finished on Valkyria Chronicles II in early 2010 , with full development beginning shortly after this . The director of Valkyria Chronicles II , Takeshi Ozawa , returned to that role for Valkyria Chronicles III . Development work took approximately one year . After the release of Valkyria Chronicles II , the staff took a look at both the popular response for the game and what they wanted to do next for the series . Like its predecessor , Valkyria Chronicles III was developed for PlayStation Portable : this was due to the team wanting to refine the mechanics created for Valkyria Chronicles II , and they had not come up with the " revolutionary " idea that would warrant a new entry for the PlayStation 3 . Speaking in an interview , it was stated that the development team considered Valkyria Chronicles III to be the series ' first true sequel : while Valkyria Chronicles II had required a large amount of trial and error during development due to the platform move , the third game gave them a chance to improve upon the best parts of Valkyria Chronicles II due to being on the same platform . In addition to Sega staff from the previous games , development work was also handled by Media.Vision. The original scenario was written Kazuki Yamanobe , while the script was written by Hiroyuki Fujii , Koichi Majima , Kishiko Miyagi , Seiki Nagakawa and Takayuki Shouji . Its story was darker and more somber than that of its predecessor .
The majority of material created for previous games , such as the BLiTZ system and the design of maps , was carried over . Alongside this , improvements were made to the game 's graphics and some elements were expanded , such as map layouts , mission structure , and the number of playable units per mission . A part of this upgrade involved creating unique polygon models for each character 's body . In order to achieve this , the cooperative elements incorporated into the second game were removed , as they took up a large portion of memory space needed for the improvements . They also adjusted the difficulty settings and ease of play so they could appeal to new players while retaining the essential components of the series ' gameplay . The newer systems were decided upon early in development . The character designs were done by Raita Honjou , who had worked on the previous Valkyria Chronicles games . When creating the Nameless Squad , Honjou was faced with the same problem he had had during the first game : the military uniforms essentially destroyed character individuality , despite him needing to create unique characters the player could identify while maintaining a sense of reality within the Valkyria Chronicles world . The main color of the Nameless was black . As with the previous Valkyria games , Valkyria Chronicles III used the CANVAS graphics engine . The anime opening was produced by Production I.G.
= = = Music = = =
The music was composed by Hitoshi Sakimoto , who had also worked on the previous Valkyria Chronicles games . When he originally heard about the project , he thought it would be a light tone similar to other Valkyria Chronicles games , but found the themes much darker than expected . An early theme he designed around his original vision of the project was rejected . He redid the main theme about seven times through the music production due to this need to reassess the game . The main theme was initially recorded using orchestra , then Sakimoto removed elements such as the guitar and bass , then adjusted the theme using a synthesizer before redoing segments such as the guitar piece on their own before incorporating them into the theme . The rejected main theme was used as a hopeful tune that played during the game 's ending . The battle themes were designed around the concept of a " modern battle " divorced from a fantasy scenario by using modern musical instruments , constructed to create a sense of atonality . While Sakimoto was most used to working with synthesized music , he felt that he needed to incorporate live instruments such as orchestra and guitar . The guitar was played by Mitsuhiro Ohta , who also arranged several of the later tracks . The game 's opening theme song , " If You Wish for ... " ( もしも君が願うのなら , Moshimo Kimi ga Negauno Nara ) , was sung by Japanese singer May 'n . Its theme was the reason soldiers fought , in particular their wish to protect what was precious to them rather than a sense of responsibility or duty . Its lyrics were written by Seiko Fujibayashi , who had worked on May 'n on previous singles .
= = = Release = = =
In September 2010 , a teaser website was revealed by Sega , hinting at a new Valkyria Chronicles game . In its September issue , Famitsu listed that Senjō no Valkyria 3 would be arriving on the PlayStation Portable . Its first public appearance was at the 2010 Tokyo Game Show ( TGS ) , where a demo was made available for journalists and attendees . During the publicity , story details were kept scant so as not to spoil too much for potential players , along with some of its content still being in flux at the time of its reveal . To promote the game and detail the story leading into the game 's events , an episodic Flash visual novel written by Fujii began release in January 2011 . The game was released January 27 , 2011 . During an interview , the development team said that the game had the capacity for downloadable content ( DLC ) , but that no plans were finalized . Multiple DLC maps , featuring additional missions and recruitable characters , were released between February and April 2011 . An expanded edition of the game , Valkyria Chronicles III Extra Edition , released on November 23 , 2011 . Packaged and sold at a lower price than the original , Extra Edition game with seven additional episodes : three new , three chosen by staff from the game 's DLC , and one made available as a pre @-@ order bonus . People who also owned the original game could transfer their save data between versions .
Unlike its two predecessors , Valkyria Chronicles III was not released in the west . According to Sega , this was due to poor sales of Valkyria Chronicles II and the general unpopularity of the PSP in the west . An unofficial fan translation patch began development in February 2012 : players with a copy of Valkyria Chronicles III